### PR TITLE
Add the to_timestamp, to_date and to_number formatting functions

### PR DESCRIPTION
**Pull request by autophagy**
## Summary of the changes / Why this improves CrateDB

This PR addresses the rest of the outstanding functions in #9663. These functions are the inverse of the behaviour of the `to_char` function, following the same pattern syntax rules (but returning their object equivalents, rather than strings).

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/io/crate/expression/scalar/ToDateFunction.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    public ToDateFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Long evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>... args) {
+        String expression = DataTypes.STRING.sanitizeValue(args[0].value());
+        String pattern = DataTypes.STRING.sanitizeValue(args[1].value());
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/ToDateFunction.java:65]:
```suggestion
        String expression = args[0].value();
        String pattern = args[1].value();
```

I think these shouldn't be necessary? They're just casts to `String` and the arguments are already `Input<String>`.

File: server/src/main/java/io/crate/expression/scalar/ToDateFunction.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    public ToDateFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Long evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>... args) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/ToDateFunction.java:63]:
```suggestion
    @SafeVarargs
    public final Long evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>... args) {
```

Fixes the warning

File: server/src/main/java/io/crate/expression/scalar/ToNumberFunction.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    public ToNumberFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Number evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>... args) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/ToNumberFunction.java:66]: See suggestions to `ToDateFunction`

File: server/src/main/java/io/crate/expression/scalar/ToTimestampFunction.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    public ToTimestampFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Long evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>... args) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/ToTimestampFunction.java:65]: See suggestions to `ToTimestampFunction`

File: docs/general/builtins/scalar.rst [hunk 5]
```diff
@@ -1290,6 +1290,267 @@ For further documentation on the behaviour of the formatter, please consult the
...
+::
+
+    TO_TIMESTAMP( expression, format_string )
+
+``expression`` is a textual representation of a timestamp.
+
+Format
+......
+
+The syntax for the format string follows that of the `Java DateTimeFormatter`_.
```
Comment (mfussenegger) [inline docs/general/builtins/scalar.rst:1314]: How compatible is this with the PostgreSQL functions?

File: docs/general/builtins/scalar.rst [hunk 6]
```diff
@@ -1290,6 +1290,267 @@ For further documentation on the behaviour of the formatter, please consult the
...
+    | ts       |
+    +----------+
+    | 63060000 |
+    +----------+
+    SELECT 1 row in set (... sec)
+
+``to_date(expression, format_string)``
+--------------------------------------
+
+The ``to_date`` function converts a textual representation of a date
```
Comment (mechanomi) [inline docs/general/builtins/scalar.rst:1411]: nit: this para might be a candidate for re-wrapping

**Review by mechanomi — COMMENTED** [review]
lgtm

**Pull request closed without merging**